### ISSUE TITLE
RSP DMA double buffering not implemented

### ISSUE THREAD

**Issue by loganmc10** [mupen64plus/mupen64plus-core#350]
As far as I can tell, no N64 emulator (PJ64, mupen64plus, mame, cen64) actually emulates double buffering in the RSP DMA's.

From my googling:
> you can have one DMA queued while another is executing

> The DMA registers are double-buffered, having one pending request and one current
active request. The DMA FULL condition means that there is an active request and
a pending request, so no more requests can be serviced

> If there is a pending DMA transaction, this transaction will be serviced before
DMA_BUSY is cleared

It sounds like it works very similarly to the AI controller. I'm curious as to why this doesn't seem to have ever really been implemented in any emulator?

To "prove" that games actually use this, if you forceably set SP_STATUS_DMA_BUSY (put ```sp->regs[SP_STATUS_REG] |= SP_STATUS_DMA_BUSY;``` in ```read_rsp_regs()``` or something), not even Mario 64 will boot. So the games are obviously checking for this. However, they don't seem to check the ```SP_DMA_BUSY_REG``` or ```SP_DMA_FULL_REG``` registers. I'm not sure why SP_STATUS_REG has bits for BUSY and FULL, and there is also a SP_DMA_BUSY_REG and SP_DMA_FULL_REG?

I was planning on working on this, but I'm just curious if anyone has any insight on why this doesn't seem to be implemented anywhere else, am I missing something?

**Comment by loganmc10** [comment]
I implemented this here:

https://github.com/loganmc10/mupen64plus-core/commit/874f4f22abdc9e78e9be46c537b6c650e37de4d8

I tested about 20 games with that commit, and the only games that no longer boot are Indiana Jones and Battle for Naboo, every other game works the same.... I don't know what these games are doing but I think I'm closing in on it...

**Comment by loganmc10** [comment]
Actually after playing around with a bit more, I realized I needed to set SP_STATUS_REG as well. Once I did that, Indiana Jones and Battle for Naboo worked again, although they don't use the FIFO queue. No game that I tested ever submitted a DMA while the DMA was busy. I guess that's why it doesn't seem to matter that this was never implemented, kind of interesting none the less. If anyone is interested what this looked like I'll leave it here:

https://github.com/loganmc10/mupen64plus-core/tree/rsp
